runtime/vm/OutOfLineINL_openj9_internal_foreign_abi_InternalDowncallHandler.cpp

```diff
@@ -216,6 +224,56 @@ OutOfLineINL_openj9_internal_foreign_abi_InternalDowncallHandler_initCifNativeTh
 	ffiTypeHelpers.freeStructFFIType(returnType);
 	goto done;
 }
+
+#if JAVA_SPEC_VERSION >= 22
+/* openj9.internal.foreign.abi.InternalDowncallHandler: private static native boolean isFfiProtoEnabled(); */
+VM_BytecodeAction
+OutOfLineINL_openj9_internal_foreign_abi_InternalDowncallHandler_isFfiProtoEnabled(J9VMThread *currentThread, J9Method *method)
+{
+	VM_BytecodeAction rc = EXECUTE_BYTECODE;
+	J9JavaVM *vm = currentThread->javaVM;
+	jboolean isFfiProtoOn = J9_ARE_ANY_BITS_SET(vm->extendedRuntimeFlags2, J9_EXTENDED_RUNTIME2_FFI_PROTO) ? JNI_TRUE : JNI_FALSE;
+
+	VM_OutOfLineINL_Helpers::returnSingle(currentThread, isFfiProtoOn, 0);
+	return rc;
+}
+
+/**
+ * openj9.internal.foreign.abi.InternalDowncallHandler: private native void resolveBoundHandle(MethodHandle handle);
+ *
+ * @brief Resolve the bounded handle via MethodHandleResolver.ffiCallLinkCallerMethod() in advance
+ *        when calling into the interpreter via MH.linkToNative()
+ *
+ * @return void
+ */
+VM_BytecodeAction
+OutOfLineINL_openj9_internal_foreign_abi_InternalDowncallHandler_resolveBoundHandle(J9VMThread *currentThread, J9Method *method)
+{
+	VM_BytecodeAction rc = EXECUTE_BYTECODE;
+	j9object_t invokeCache = NULL;
+	j9object_t nativeMH = J9_JNI_UNWRAP_REFERENCE(currentThread->sp);
+
+	/* The resolution is performed by MethodHandleResolver.ffiCallLinkCallerMethod()
+	 * to obtain a MemberName object plus appendix in a two element object array.
+	 */
+	VM_OutOfLineINL_Helpers::buildInternalNativeStackFrame(currentThread, method);
+	VM_VMHelpers::pushObjectInSpecialFrame(currentThread, nativeMH);
+	invokeCache = resolveFfiCallInvokeHandle(currentThread, nativeMH);
+	VM_VMHelpers::popObjectInSpecialFrame(currentThread);
+	if (VM_VMHelpers::exceptionPending(currentThread)) {
+		rc = GOTO_THROW_CURRENT_EXCEPTION;
+		goto done;
+	}
+
```

Comment:
Does `nativeMH` need to be refreshed here?

Reply:
There is no need to do so as the intention of `push/popObjectInSpecialFrame`  is to save the object across the GC point so as to prevent GC from updating the object when resolving the the bound MH.

Reply:
I think it does need to be refreshed:
```c
	nativeMH = VM_VMHelpers::popObjectInSpecialFrame(currentThread);
```
See also https://github.com/eclipse-openj9/openj9/blob/master/runtime/vm/BytecodeInterpreter.hpp#L3366-L3374

@gacholio Can you offer your opinion here?

Reply:
Yes we should reset `nativeMH` from the returned object `VM_VMHelpers::popObjectInSpecialFrame(currentThread);` (I previously misunderstood your comments). I've updated the code there.

Reply:
Yes, it's needed even if you aren't going to use it again (the stack has to be popped in all cases).

Reply:
The pop was there before, it was the re-assignment to `nativeMH` that I thought was missing.

Reply:
Yes. I already updated the code to reassigned the return value to `nativeMH`.

Reply:
@gacholio , Keith's concern is not the pop operation itself (which was already been there) but the update for `nativeMH`. I recalled that the object reference is kept unchanged with `push/popObjectInSpecialFrame` across the GC point. If so, it should be fine not to update the `nativeMH` with the return value from `popObjectInSpecialFrame` (I personally tried this way which seems working)

Reply:
As confirmed by GAC offline, `nativeMH` must be reassigned given it is used to set the invoke cache before returning to  java in the native code.